### PR TITLE
Remove unused includes from executor/multi_utility.c

### PR DESCRIPTION
**Pull request by lithp**
The sheer number of unnecessary includes in this file and many others makes me think we're not removing them for a reason. But I don't know what it is, so here's a PR :)

### CODE REVIEW

File: src/backend/distributed/executor/multi_utility.c [hunk 1]
```diff
@@ -14,61 +14,32 @@
 
 #include <string.h>
 
-#include "access/attnum.h"
```
Comment (jasonmp85) [inline src/backend/distributed/executor/multi_utility.c]: Provides `InvalidAttrNumber`.

File: src/backend/distributed/executor/multi_utility.c [hunk 2]
```diff
@@ -14,61 +14,32 @@
...
-#include "access/heapam.h"
```
Comment (jasonmp85) [inline src/backend/distributed/executor/multi_utility.c]: Provides `heap_openrv`, others.

File: src/backend/distributed/executor/multi_utility.c [hunk 3]
```diff
@@ -14,61 +14,32 @@
...
-#include "access/htup.h"
```
Comment (jasonmp85) [inline src/backend/distributed/executor/multi_utility.c]: Provides `HeapTuple`.

File: src/backend/distributed/executor/multi_utility.c [hunk 4]
```diff
@@ -14,61 +14,32 @@
...
 #include "access/htup_details.h"
 #include "access/sysattr.h"
-#include "access/tupdesc.h"
```
Comment (jasonmp85) [inline src/backend/distributed/executor/multi_utility.c]: Provides `TupleDesc`

File: src/backend/distributed/executor/multi_utility.c [hunk 5]
```diff
@@ -14,61 +14,32 @@
...
-#include "access/xact.h"
```
Comment (jasonmp85) [inline src/backend/distributed/executor/multi_utility.c]: Provides `CommandCounterIncrement`

**Comment by jasonmp85** [comment]
Hey, so as far as I can see, the _only_ `#include` that is actually unused is `utils/guc.h`. Everything else is used:

```
#include <string.h>                               // for NULL, strcmp, strncmp
#include "access/attnum.h"                        // for InvalidAttrNumber
#include "access/heapam.h"                        // for heap_openrv, heap_c...
#include "access/htup.h"                          // for HeapTuple, HeapTupl...
#include "access/htup_details.h"                  // for GETSTRUCT
#include "access/sysattr.h"                       // for FirstLowInvalidHeap...
#include "access/tupdesc.h"                       // for tupleDesc, TupleDesc
#include "access/xact.h"                          // for CommandCounterIncre...
#include "c.h"                                    // for bool, Oid, true, false
#include "catalog/catalog.h"                      // for IsSystemClass
#include "catalog/dependency.h"                   // for sequenceIsOwned
#include "catalog/index.h"                        // for IndexGetRelation
#include "catalog/namespace.h"                    // for RangeVarGetRelid
#include "catalog/pg_attribute.h"                 // for Form_pg_attribute
#include "catalog/pg_class.h"                     // for FormData_pg_class
#include "commands/defrem.h"                      // for defGetQualifiedName
#include "commands/tablecmds.h"                   // for AlterTableGetLockLevel
#include "distributed/citus_ruleutils.h"          // for generate_relation_name
#include "distributed/commit_protocol.h"          // for TransactionConnection
#include "distributed/connection_cache.h"         // for ConnectionGetOption...
#include "distributed/master_metadata_utility.h"  // for ShardInterval, Load...
#include "distributed/master_protocol.h"          // for WORKER_APPLY_SHARD_...
#include "distributed/metadata_cache.h"           // for IsDistributedTable
#include "distributed/multi_copy.h"               // for NodeAddress, CitusC...
#include "distributed/multi_join_order.h"         // for PartitionKey, Parti...
#include "distributed/multi_shard_transaction.h"  // for ShardConnections
#include "distributed/multi_utility.h"
#include "distributed/pg_dist_partition.h"        // for DISTRIBUTE_BY_APPEND
#include "distributed/resource_lock.h"            // for LockShards
#include "distributed/transmit.h"                 // for FreeStringInfo, Red...
#include "distributed/worker_protocol.h"          // for CacheDirectoryElement
#include "executor/executor.h"                    // for ExecCheckRTPerms
#include "foreign/foreign.h"                      // for GetForeignServerByName
#include "lib/stringinfo.h"                       // for StringInfoData, app...
#include "libpq-fe.h"                             // for PQclear, PQexec
#include "miscadmin.h"                            // for GetUserId, SetUserI...
#include "nodes/bitmapset.h"                      // for bms_add_member
#include "nodes/nodes.h"                          // for Node, IsA, makeNode
#include "nodes/params.h"                         // for ParamListInfo
#include "nodes/parsenodes.h"                     // for CopyStmt, IndexStmt
#include "nodes/pg_list.h"                        // for List, ListCell, Lis...
#include "nodes/primnodes.h"                      // for RangeVar, Var, OnCo...
#include "nodes/value.h"                          // for ValUnion, strVal
#include "port.h"                                 // for make_absolute_path
#include "postgres.h"                             // for Datum, ExceptionalC...
#include "storage/lmgr.h"                         // for LockRelationOid
#include "storage/lock.h"                         // for LOCKMODE, NoLock
#include "tcop/dest.h"                            // for DestReceiver, None_...
#include "tcop/utility.h"                         // for ::PROCESS_UTILITY_T...
#include "utils/acl.h"                            // for aclcheck_error, pg_...
#include "utils/builtins.h"                       // for quote_literal_cstr
#include "utils/elog.h"                           // for errmsg, errcode
#include "utils/errcodes.h"                       // for ERRCODE_FEATURE_NOT...
#include "utils/hsearch.h"                        // for HTAB
#include "utils/inval.h"                          // for AcceptInvalidationM...
#include "utils/lsyscache.h"                      // for get_namespace_name
#include "utils/memutils.h"                       // for TopTransactionContext
#include "utils/palloc.h"                         // for MemoryContextSwitchTo
#include "utils/rel.h"                            // for RelationData, Relat...
#include "utils/relcache.h"                       // for Relation
#include "utils/syscache.h"                       // for ReleaseSysCache
```

What is your definition of _used_? Is it "if we remove it, it still compiles"? What cost do the extra includes incur?

**Comment by jasonmp85** [comment]
(IMO, except in edge cases, omitting a header because some other header transitively provides the symbol you need should be an error).

**Comment by anarazel** [comment]
> (IMO, except in edge cases, omitting a header because some other header transitively provides the symbol you need should be an error).

I still think the only benefit of that position is making refactoring of
include structures harder and requiring pointless tool hint comments.

**Comment by jasonmp85** [comment]
> the only benefit of that position is making refactoring of include structures harder

Except this cuts both ways: what if you want to remove an `#include` someone else transitively depends on? Why should you have to modify unrelated files because you've removed your internal need for a particular `#include`? It violates locality?

**Comment by anarazel** [comment]
On 2016-08-19 13:31:55 -0700, Jason Petersen wrote:
> > the only benefit of that position is making refactoring of include structures harder
> 
> Except this cuts both ways: what if you want to remove an `#include` someone else transitively depends on? Why should you have to modify unrelated files because you've removed your internal need for a particular `#include`? It violates locality?

It doesn't help there either - a new file won't be included by IWYU type
logic.

**Comment by jasonmp85** [comment]
I mean the refactor where `a.c` includes `b.h` which includes `c.h`.

Assume `a.c` needs symbols defined in both `b.h` and `c.h`, but only includes `b.h`.

Someone changes `b.h` to remove the dependency on `c.h`. Now `a.c` fails to compile.

If `a.c` had correctly enumerated its dependencies, this would not have happened and the person doing the refactor would not need to change the unrelated and otherwise unaffected file `a.c` in their patch.

In fact, in the above case, `iwyu` would suggest to the original author of `a.c` that they include `c.h` as they have an explicit dependency on a symbol from it.

**Comment by jasonmp85** [comment]
@lithp — How did you produce this patch? Manually? Or is there another tool that does the opposite of `iwyu`?

I mostly view the `#includes` section as a big section at the top I skip over when reading. It's there if I want to see dependency analysis to find who actually depends on what files, but largely it's for the computer, not me.

I think there's more the computer can do if we enumerate all explicit symbol dependencies, but honestly my major gripe is I want a tool to do this for me rather than having to hand-curate a list of "where to find stuff" for the compiler.

So if you have another tool (that reduces `include`s as much as possible?) let me know and maybe we can sidestep this discussion with some sort of automation.

**Comment by lithp** [comment]
Heh, sorry, didn't mean to start anything here. I've been primed by some python tooling to be suspicious of long include lists and to try to minimize them, but python doesn't have any equivalent of indirect includes.

Since the things to be included aren't enumerated I think both forms aren't very readable so minimizing unrelated changes in commits probably takes precedence. 

> I think there's more the computer can do if we enumerate all explicit symbol dependencies

Is this actually possible in c? Do you mean forward declaring everything we use?

> So if you have another tool (that reduces includes as much as possible?) let me know and maybe we can sidestep this discussion with some sort of automation.

I did this manually and was going to propose making a tool for it but the internet has pointed out lots of edge cases and made me wary about doing so.

**Pull request closed without merging**